### PR TITLE
Game_logic

### PR DESCRIPTION
**Pull request by BigWizzo**
# Ruby: OOP - Milestone: game logic

> Using ruby object-oriented programming implementing the tic-tac-toe game_logic.
> Encapsulating different parts of the game logic in classes.

### CODE REVIEW

File: bin/main
Comment (teekaytech) [inline bin/main]: Remove all the inline comments in this file and others.

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__This is an awesome project. Kudos for including the README.md in this milestone :clap: However, some changes must be made before this project can be approved. Please check my comments below and make the changes as appropriate.__

**[ REQUIRED CHANGES ]**

- [x] It looks like you dump the whole of your README.md file content as the summary of your Pull Request. You do not have to do this. You only need to add some statements that summarize what you've done in this milestone. Kindly change the summary of your PR.

- [x] In this milestone, there should be no inline comments left after the previous milestone. 
([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/no_comment.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [x] Your app is currently allowing a player to move into an occupied position. This should not be the case. Kindly fix this bug.



- [x] There seems to be a bug in the way your app checks the winning move. Consider the 2 instances below:

**Instance-1**



**Instance-2**


Please study the images, and use it to fix how your game is checking the winning moves.

- [x] Also, your draw logic seems to be buggy. It announces a game as a draw when all the moves have not been made. The game should only announce a draw when all the available moves have been made and no winning move is made. See the image below:



* * * 

**Please make the required changes and submit another review request when you are done.**

* * *

**_You can reach out to me on Twitter (@ola_lere), Github(@teekaytech), or slack (@taofeekolalere )_**
__Happy Coding :smiley:__

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request reviewed by Maria Eugenia Queme P. (TSE)

### Hello @BigWizzo and @Gabkings :lollipop: :balloon: 

### Thank you for working on the changes requested by the previous reviewer :medal_sports: :medal_sports: I have to say I always like when tic-tac-toe games include detailed instructions and also request for the player's names I think this makes it stand a lot from the rest :video_game: :star: also claps to you for having such a structured Readme file at this milestone :clap: :clap: :clap:  let's work on the following changes :construction: :construction_worker: :construction_worker: 

#### Behaviour

- [x] The game is letting players have empty names
> let's validate it so either it requests for the name again or assigns a name by default like 'P1' and 'P2' for example.
image

- [x] Even when the names of the players are given they aren't showing in the game when it's needed
> for instance in the instructions it should say "Maru you play first your chip is X" but instead only says "you play first your chip is X" let's fix this so the players' name shows whenever something is required from them
image

- [x] The game should render the first board (only numbers) at the beginning of the match
image

- [x] The application does not allow a player to move into an occupied position
> even though the game doesn't allow it the right message should appear right now this message "Your input is invalid. Put an available number between 1 - 9 on unoccupied spot" is used whenever an invalid input is given or an already occupied position is given and when that's the case it should say something like "This position has already been occupied" 
image

#### Implementation details

* [x] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
> Let's make more class methods and take some responsibility from the main.RB file (all game validations should be handled by methods) the reason I insist on this it's because later on, you'll build test cases for this project.


### That's it, team! :raised_hands: 

### Please :pray: do the changes :technologist: :technologist: and submit for a code review when you are done :memo: 

### Happy coding! :rocket: 

Any questions? ask me on slack **@Maru** :technologist: 
PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

File: bin/main
Comment (descholar-ceo) [inline bin/main]: Remember to change this into `require_relative '../lib/game_logic.rb'`, because there isn't `logic.rb` in your lib folder, but `game_logic.rb` instead

File: lib/game_logic.rb
Comment (descholar-ceo) [inline lib/game_logic.rb]: This method is returning an undefined `return_value`. Advice: Initialize `return_value` out of the loop, and your codes seem that they should be changing the value of that `return_value` inside the loop :point_right: line 44, so that you will be able to return it on line 46.

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Change Required :vertical_traffic_light: :warning: 
## Hello @BigWizzo and @Gabkings  :handshake: :handshake: :handshake: 
:+1: This is good to make it this far, your project is working almost as it should, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
## Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes]
- [x] Your project is crashing upon running it, that's because you are requiring an unexistant file from the lib folder, please read the inline comments I added in directly in your main file and fix that issue
- [x]  Again your project is crashing upon playing it, read the error message on the screenshot below
Screenshot from 2020-10-27 21-51-20 :point_right: I suggested some changes that you can add in your method to make it work
- [x] Your `wining` and `drawing` still have some bug that needs to be worked on, when I played the game, it's showing that's a draw whereas it was a winning. After the winning move, you should stop the game and ask players if they need to play again or not, so that they choose to continue or terminate the game. Look at the below screenshots:
1. This move is a winning move, and the game should ask players wether they want to play again or not, 
Screenshot from 2020-10-27 21-57-54

2. As you can see, the winning logic is not woring at all, because it is not logical to have two winners on the same game round
Screenshot from 2020-10-27 21-58-41

3. The drawing logic seems to me that it has some bug as well, because when the whole board is played, it is showing that is a draw, 
Screenshot from 2020-10-27 22-00-23
- [x] Implement all changes I suggested in inline comments in your codes

# [Optional changes]
- [x] As you displayed the board with numbers on the beginning of the game, you should continue displaying numbers even while they are playing, because it gives a good user UI/UX making it easy for players to find easily box to play.
#### Note: About this, don't be confused with the previous reviewer suggestion ` The game should render the first board (only numbers) at the beginning of the match`, they were talking about when the game is first loaded, they were not referring to when the game is being played.

- [x] Your commit messages should be in imperative mode or present tense, you should have used `Add board display logic` instead of `Adding board display logic`, refer [to this article for more details on how to write a good && professional commit message](https://www.freecodecamp.org/news/level-you-up-to-awesome-commit-messages-a85558cb90e8)

If anything seems to be unclear to you, don't hesitate to contact me for more clarification slack: `@Mugirase Emmanuel` | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)

**Review by ElijahTobs — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: :red_circle: 

## Hi @BigWizzo  and @Gabkings 

:+1:   I can see you've put in a lot of work and I commend you for that
:+1:  Although you've worked on some of the required changes by the previous reviewers, there's still some work to do before this project can be approved.

## [MANDATORY]

- [x] You need to work on your winning logic. The screenshot below shows that Elijah already won with that move which should end the game and declare Elijah a winner but that didn't happen.
e1

- [x] The second player also experiences the same thing and it's not even possible to have 2 winners
e2

- [x] The draw logic is also not left out. There are two possible winners, yet it shows it's a draw
e3

## [OPTIONAL]

Your game will look better and will be easier to play/understand if you continue showing players the corresponding number of available slots on the board
e5

### Effect the changes and submit for another review
### Happy Coding :muscle: :computer: 

Feel free to reach out to me on slack **`@Elijah Ayandokun`**
   **Elijah O. Ayandokun**
[Github](https://github.com/ElijahTobs) | [LinkedIn](https://www.linkedin.com/in/ayandokunelijah/) | [Twitter](https://twitter.com/ElijahTobs)

**Review by arslanbisharat — APPROVED** [review]
# Approved :smile: 

Hello Developer, :wave:

I just reviwed your project and I am very happy to annocunce that your project is approved.
Congraulation for achieving this milestone towards your career. :clap:

_I am looking forward towards to see your next project._

-----------------------------------------------------------------------------------------------

#### Regards,

#### Arslan Bisharat

#### slack: arslanbisharat

**Pull request merged**
